File: day03/day03.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Serilog;

namespace Shunty.AdventOfCode2019
{
    public class Day03 : IAoCRunner
    {
        private static readonly int DayNumber = 3;

        public void Run(ILogger log)
        {
            var input = AocHelpers.GetDayLines(DayNumber)
                .Select(l => l.Split(',').ToList())
                .ToList();

            var path1 = MapPath(input[0]);
            var path2 = MapPath(input[1]);

            var intersection = path1       // Could use .Intersect() here with a custom
                .Join(                     // IEqualityComparer<Location>. But why bother
                    path2,                 // when .Join() works fine.
                    p => (p.X, p.Y),
                    p => (p.X, p.Y),
                    (p1, p2) => new { X = p1.X, Y = p1.Y, Distance = p1.Distance, Steps = p1.Steps + p2.Steps }
                );

            var part1 = intersection.OrderBy(p => p.Distance).First();
            Console.WriteLine($"Part 1: {part1.Distance} (X={part1.X}; Y={part1.Y})");
            var part2 = intersection.OrderBy(p => p.Steps).First();
            Console.WriteLine($"Part 2: {part2.Steps} (X={part2.X}; Y={part2.Y})");
        }

        internal struct Location
        {
            public int X;
            public int Y;
            public int Steps;
            public int Distance => Math.Abs(X) + Math.Abs(Y);
        }

        private List<Location> MapPath(IEnumerable<string> path)
        {
            // Walk through each instruction and add a location to the result for every
            // grid point visited, including number of steps to get there.
            var result = new List<Location>();
            int cx = 0, cy = 0, steps = 0;
            foreach (var inst in path)
            {
                var direction = inst[0];
                var distance = int.Parse(inst.Substring(1));
                foreach (var d in Enumerable.Range(1, distance))
                {
                    switch (direction)
                    {
                        case 'D':
                            cy--;
                            break;
                        case 'U':
                            cy++;
                            break;
                        case 'L':
                            cx--;
                            break;
                        case 'R':
                            cx++;
                            break;
                        default:
                            throw new Exception($"Unknown direction \"{direction}\"");
                    }
                    result.Add(new Location { X = cx, Y = cy, Steps = ++steps });
                }
            }
            return result;
        }
    }
}

File: day03/day03.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"strconv"
	"strings"
)

func main() {
	input, err := getInput("./day03-input.txt")
	if err != nil {
		log.Fatal(err)
	}

	path1 := plotPath(strings.Split(input[0], ","))
	path2 := plotPath(strings.Split(input[1], ","))
	part1, part2 := math.MaxInt32, math.MaxInt32

	for gr, steps1 := range path1 {
		if steps2, ok := path2[gr]; ok {
			// Part 1 - Find the closest (Manhattan distance) common point
			dist := int(math.Abs(float64(gr.X)) + math.Abs(float64(gr.Y)))
			if dist < part1 {
				part1 = dist
			}
			// Part 2 - Find the shortest total distance to a common point
			totalSteps := steps1 + steps2
			if totalSteps < part2 {
				part2 = totalSteps
			}
		}
	}
	fmt.Println("Part 1: ", part1)
	fmt.Println("Part 2: ", part2)
}

// GridRef : A 2D grid reference
type GridRef struct {
	X, Y int
}

func plotPath(path []string) map[GridRef]int {
	result := make(map[GridRef]int)
	x, y, dx, dy := 0, 0, 0, 0
	steps := 0
	for _, instruction := range path {
		if instruction == "" {
			continue
		}
		dir := string(instruction[0])
		dist, _ := strconv.Atoi(instruction[1:])
		switch dir {
		case "R":
			dx, dy = 1, 0
		case "L":
			dx, dy = -1, 0
		case "U":
			dx, dy = 0, 1
		case "D":
			dx, dy = 0, -1
		default:
			// Error
		}

		for d := 0; d < dist; d++ {
			steps++
			x += dx
			y += dy
			result[GridRef{x, y}] = steps
		}
	}
	return result
}

func getInput(inputFile string) ([]string, error) {
	txt, err := ioutil.ReadFile(inputFile)
	if err != nil {
		return nil, err
	}
	return strings.Split(string(txt), "\n"), nil
}
